### PR TITLE
Feature/issue 216

### PR DESCRIPTION
**Pull request by marcfiu**
Draft PR for linode.cloud.firewall `state=update` feature. The focus of the `update` feature was for rules. The intent is for a workflow such that one can update specifically labeled rules without needing to know their position within the rules list. For example, a user may add arbitrary rules to a firewall (manually or using some other tool) and then one wants to use ansible to update specifically labeled rules.

### CODE REVIEW

File: tests/integration/targets/firewall_update/tasks/main.yaml
Comment (marcfiu) [inline tests/integration/targets/firewall_update/tasks/main.yaml:1]: @LBGarber: I don't know how you run your test suite. I ran this manually separately with my own harness. For this reason, whatever I put here might not work out of the box with whatever test suite you use.

Comment (lgarber-akamai) [reply tests/integration/targets/firewall_update/tasks/main.yaml:1]: We generally run the following command to execute the e2e test suite:
```
make TEST_ARGS="-v test_target_name" test
```

Fortunately this looks good for the most part, so it shouldn't be too much trouble to get it compatible.

File: plugins/modules/firewall.py
Comment (lgarber-akamai) [inline plugins/modules/firewall.py:245]: Small return type issue here
```suggestion
    def _amend_rules(self, remote_rules: list, local_rules: list) -> list:
```

File: plugins/modules/firewall.py
Comment (lgarber-akamai) [inline plugins/modules/firewall.py:265]: It might be a bit cleaner to return an `Optional[dict]` here since we can return `None` when no rules need to be updated and the structure defined [here](https://www.linode.com/docs/api/networking/#firewall-rules-update__request-body-schema) if the rules need to be updated.

**Review by lgarber-akamai — CHANGES_REQUESTED** [review]
Looks great so far! Just a few things I came across :slightly_smiling_face:

File: plugins/modules/firewall.py
Comment (lgarber-akamai) [inline plugins/modules/firewall.py:278]: I wonder if there's a way we could break this out into another function or use guard clauses to make this a bit more readable. No big deal either way since it's primarily a preference and formatting thing :slightly_smiling_face:

File: plugins/modules/firewall.py
Comment (marcfiu) [reply plugins/modules/firewall.py:265]: Do you mean for `_update_rules()` or the other method above? I'm guessing the former. Right now it returns an empty list if there are no changes, or a list of changes.  I do not believe it ever returns None.

Comment (marcfiu) [reply plugins/modules/firewall.py:265]: Need clarification wrt the structured defined in the REST API. Previously the changed_rules() method this is based on returned a boolean and only upon True would the caller then actually update things. For this reason, maybe if this ever was supposed to return None to nullify a firewall it wasn't done that before either?

Comment (lgarber-akamai) [reply plugins/modules/firewall.py:265]: Sorry for the late response!

> Do you mean for _update_rules() or the other method above? I'm guessing the former. Right now it returns an empty list if there are no changes, or a list of changes. I do not believe it ever returns None.

In this case I'm thinking more of a structural change. Since the [Firewall Rules Update](https://www.linode.com/docs/api/networking/#firewall-rules-update) endpoint accepts a dict describing the entire configuration of firewall rules (`inbound`, `inbound_policy`, `outbound`, and `outbound_policy`), I'm thinking it might be ideal for us to return `None` if there are no changes to be made and return the complete update request dict otherwise.

File: plugins/modules/firewall.py
Comment (marcfiu) [reply plugins/modules/firewall.py]: I opted not to do that for now.

File: plugins/modules/firewall.py
Comment (marcfiu) [reply plugins/modules/firewall.py]: _update_rules() always returns a dict, even an empty one.

_change_rules() will either return the new rules Dict or None if there are no changes.

Does this address your comment?

**Review by lgarber-akamai — APPROVED** [review]
Looks good to me! Waiting on a second review and we'll be good to merge

**Pull request merged**
